### ISSUE TITLE
Wrong netty version

### ISSUE THREAD

**Issue by frank-montyne** [elastic/elasticsearch#14744]
Hi,

I get the following error when using the elasticsearch 2.0 java client libraries:

**java.lang.NoClassDefFoundError: org/jboss/netty/util/internal/StackTraceSimplifier**

Elasticsearch 2.0 ships with the "netty-3.10.5.Final.jar" that should contain that class but it doesn't. Only much older versions of the netty jar like the "netty-3.2.5.Final.jar" contain that class.

Is this a problem in the distribution? I didn't test this with earlier versions but I thinks they have a similar problem.

Below is the stacktrace: I'm running my persistence service (that connects to elasticsearch) as a linux service using JAJSW (yet another java service wrapper)

INFO|3861/0|15-11-13 12:55:55|java.lang.reflect.InvocationTargetException
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
INFO|3861/0|15-11-13 12:55:55|  at java.lang.reflect.Method.invoke(Method.java:497)
INFO|3861/0|15-11-13 12:55:55|  at org.rzo.yajsw.app.WrapperJVMMain.executeMain(WrapperJVMMain.java:53)
INFO|3861/0|15-11-13 12:55:55|  at org.rzo.yajsw.app.WrapperJVMMain.main(WrapperJVMMain.java:36)
**INFO|3861/0|15-11-13 12:55:55|Caused by: java.lang.NoClassDefFoundError: org/jboss/netty/util/internal/StackTraceSimplifier**
INFO|3861/0|15-11-13 12:55:55|  at org.jboss.netty.logging.InternalLoggerFactory.<clinit>(InternalLoggerFactory.java:49)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.netty.NettyUtils.<clinit>(NettyUtils.java:90)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.transport.netty.NettyTransport.<clinit>(NettyTransport.java:106)
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
INFO|3861/0|15-11-13 12:55:55|  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
INFO|3861/0|15-11-13 12:55:55|  at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:56)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:86)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:104)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.FactoryProxy.get(FactoryProxy.java:54)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:47)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:865)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:59)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:46)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.SingleParameterInjector.inject(SingleParameterInjector.java:42)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.SingleParameterInjector.getAll(SingleParameterInjector.java:66)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:85)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:104)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:47)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:865)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:59)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:46)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:201)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:193)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:858)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorBuilder.loadEagerSingletons(InjectorBuilder.java:193)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorBuilder.injectDynamically(InjectorBuilder.java:175)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:110)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:93)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:70)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.common.inject.ModulesBuilder.createInjector(ModulesBuilder.java:47)
INFO|3861/0|15-11-13 12:55:55|  at org.elasticsearch.client.transport.TransportClient$Builder.build(TransportClient.java:147)
INFO|3861/0|15-11-13 12:55:55|  at com.bynubian.elasticsearch.base.ElasticSearchDB.connect(ElasticSearchDB.java:219)
INFO|3861/0|15-11-13 12:55:55|  at com.bynubian.elasticsearch.base.ElasticSearchDB.connect(ElasticSearchDB.java:147)
INFO|3861/0|15-11-13 12:55:55|  at com.bynubian.services.persistence.server.PersistenceServer.initialize(PersistenceServer.java:92)
INFO|3861/0|15-11-13 12:55:55|  at be.landc.framework.service.server.shared.Server.initialize(Server.java:98)
INFO|3861/0|15-11-13 12:55:55|  at be.landc.framework.service.ipl.Boot.initialize(Boot.java:211)
INFO|3861/0|15-11-13 12:55:55|  at be.landc.framework.service.ipl.Boot.<init>(Boot.java:82)
INFO|3861/0|15-11-13 12:55:55|  at be.landc.framework.service.ipl.Boot.main(Boot.java:320)
INFO|3861/0|15-11-13 12:55:55|  ... 6 more
INFO|3861/0|15-11-13 12:55:55|Caused by: java.lang.ClassNotFoundException: org.jboss.netty.util.internal.StackTraceSimplifier
INFO|3861/0|15-11-13 12:55:55|  at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
INFO|3861/0|15-11-13 12:55:55|  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
INFO|3861/0|15-11-13 12:55:55|  at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
INFO|3861/0|15-11-13 12:55:55|  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
INFO|3861/0|15-11-13 12:55:55|  ... 48 more

**Comment by rmuir** [comment]
Possibly jar hell in your setup. If you are using elasticsearch 2.0 as a client, first try this:
```
import org.elasticsearch.bootstrap.JarHell;

...

    JarHell.checkJarHell();
```

It might shed some light on the situation..

**Comment by frank-montyne** [comment]
Thanks I'll give that a try

**Comment by frank-montyne** [comment]
Running JarHell doesn't indicate that something is wrong

**Comment by clintongormley** [comment]
Isn't the problem in JBoss?

    INFO|3861/0|15-11-13 12:55:55| at org.jboss.netty.logging.InternalLoggerFactory.(InternalLoggerFactory.java:49)
    INFO|3861/0|15-11-13 12:55:55| at org.elasticsearch.common.netty.NettyUtils.(NettyUtils.java:90)

**Comment by jasontedor** [comment]
> Isn't the problem in JBoss?

Netty use to be developed under JBoss so the package names from the 3.x line reflect that (it is no the longer case).

I think that the user has something unusual going on with Netty and their classpath. In particular, the `StackTraceSimplifier` class that you refer to was removed between 3.9.0 and 3.10.0 of Netty. If you look at the source of `InternalLoggerFactory` from say 3.9.0.Final, you'll see that it is used in the class initializer for `InternalLoggerFactory` (and this is not the case in 3.10.5.Final). If that class is not available, that class initializer for `InternalLoggerFactory` fails and that leads to a [`NoClassDefFoundError`](http://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.4.2). The version of Netty that Elasticsearch depends on (3.10.5.Final) should not be looking for StackTraceSimplifier because it has been completely removed from Netty in that version.

I took a cursory peek at YAJSW, and it appears to have a Netty dependency (maybe Netty 3.6.6.Final?). I suspect that this is the underlying source of the issue.
